src/ReservoirPriceOracle.sol

```diff
@@ -353,40 +456,87 @@ contract ReservoirPriceOracle is IPriceOracle, IReservoirPriceOracle, Owned(msg.
+    function setRoute(address aToken0, address aToken1, address[] memory aRoute) public onlyOwner {
         if (aToken0 == aToken1) revert RPC_SAME_TOKEN();
         if (aToken1 < aToken0) revert RPC_TOKENS_UNSORTED();
         if (aRoute.length > MAX_ROUTE_LENGTH || aRoute.length < 2) revert RPC_INVALID_ROUTE_LENGTH();
         if (aRoute[0] != aToken0 || aRoute[aRoute.length - 1] != aToken1) revert RPC_INVALID_ROUTE();
 
-        _route[aToken0][aToken1] = aRoute;
-        emit Route(aToken0, aToken1, aRoute);
-
-        // iteratively define those subroutes if undefined
-        if (aRoute.length > 2) {
-            for (uint256 i = 0; i < aRoute.length - 1; ++i) {
-                (address lLowerToken, address lHigherToken) = aRoute[i].sortTokens(aRoute[i + 1]);
+        bytes32 lSlot = _calculateSlot(aToken0, aToken1);
 
-                // if route is undefined
-                address[] memory lExisting = _route[lLowerToken][lHigherToken];
-                if (lExisting.length == 0) {
-                    address[] memory lSubroute = new address[](2);
-                    lSubroute[0] = lLowerToken;
-                    lSubroute[1] = lHigherToken;
+        // simple route
+        // gas: can store aRoute.length as a local variable
+        if (aRoute.length == 2) {
+            assembly {
+                let data := shl(248, 0x01) // 0x01 in the uppermost byte
+                sstore(lSlot, data)
+            }
+            // update price for simple routes
+        }
+        // composite route
+        else {
+            uint256 lIndex = 0;
+
+            // first + intermediate hops
+            for (uint256 i = 1; i < aRoute.length - 1; ++i) {
+                address lNextToken = aRoute[i];
+                assembly {
+                    let data := shl(248, 0x02) // 0x02 in the uppermost byte
+                    lNextToken := shl(88, lNextToken)
+                    data := or(data, lNextToken)
+                    sstore(add(lSlot, lIndex), data) // need to do funny things with index?
+                    lIndex := add(lIndex, 1)
+                }
 
-                    _route[lLowerToken][lHigherToken] = lSubroute;
-                    emit Route(lLowerToken, lHigherToken, lSubroute);
+                (address lLowerToken, address lHigherToken) = aRoute[i - 1].sortTokens(aRoute[i]);
+                bytes32 lIntermediateRouteSlot = _calculateSlot(lLowerToken, lHigherToken);
+                bytes32 lRead;
+                assembly {
+                    lRead := sload(lIntermediateRouteSlot)
+                }
+                if (lRead == bytes32(0)) {
+                    address[] memory lIntermediateRoute = new address[](2);
+                    lIntermediateRoute[0] = lLowerToken;
+                    lIntermediateRoute[1] = lHigherToken;
+                    setRoute(lLowerToken, lHigherToken, lIntermediateRoute);
                 }
             }
```

Comment:
Please comment what is happening here, its very hard to follow

Reply:
refactored and added comments